client/asset/dcr/dcr.go

```diff
@@ -1912,29 +1930,38 @@ func (dcr *ExchangeWallet) fund(keep uint64, // leave utxos for this reserve amt
 		return nil, nil, 0, 0, asset.ErrInsufficientBalance
 	}
 
-	sum, extra, sz, coins, spents, redeemScripts, err := tryFund(utxos, enough)
-	if err != nil {
-		return nil, nil, 0, 0, err
-	}
-
-	if avail-sum+extra < keep {
-		if extra != 0 { // there was usable change from enough, just not enough
-			return nil, nil, 0, 0, asset.ErrInsufficientBalance
-		} // else caller probably not granting change
-		// NOTE: tryFund above may not have tried 0-conf, but we may now.
+	var sz uint32
+	var spents []*fundingCoin
 
-		dcr.log.Debugf("Trying again with some UTXOs excluded for required reserves of %v DCR",
-			toDCR(keep))
-
-		// Try again with some utxos taken out of the mix for keep. Select these
-		// with the objective of being as close to the amount as possible,
-		// unlike tryFund that minimizes the number of UTXOs chosen.
+	// First take some UTXOs out of the mix for any keep amount. Select these
+	// with the objective of being as close to the amount as possible, unlike
+	// tryFund that minimizes the number of UTXOs chosen. By doing this first,
+	// we may be making the order spend a larger number of UTXOs, but we
+	// mitigate subsequent order funding failure due to reserves because we know
+	// this order will leave behind sufficient UTXOs without relying on change.
+	if keep > 0 {
 		kept := leastOverFund(keep, utxos)
-		utxos = utxoSetDiff(utxos, kept)
-		sum, _, sz, coins, spents, redeemScripts, err = tryFund(utxos, enough)
-		if err != nil { // no joy with the reduced set
+		dcr.log.Debugf("Setting aside %v DCR in %d UTXOs to respect the %v DCR reserved amount",
+			toDCR(sumUTXOs(kept)), len(kept), toDCR(keep))
+		utxosPruned := utxoSetDiff(utxos, kept)
+		sum, _, sz, coins, spents, redeemScripts, err = tryFund(utxosPruned, enough)
+		if err != nil { // try with the full set
+			dcr.log.Debug("Unable to fund order with UTXOs set aside (%v), trying again with full UTXO set.", err)
+		} // else spents is populated
+	}
+	if len(spents) == 0 { // either keep is zero or it failed with utxosPruned
+		// Without utxos set aside for keep, we have to consider any spendable
+		// change (extra) that the enough func grants us.
+		var extra uint64
+		sum, extra, sz, coins, spents, redeemScripts, err = tryFund(utxos, enough)
```

Comment:
This commit also reverses the order of funding attempts vs the initial commit that updating `fund`/`tryFund`.  It tries straight away to set aside UTXOs for the reserves, and falls back to the standard funding approach, hoping that the enough func grants usable change in the extra output.